simplified-app-shared/src/main/java/org/nypl/simplified/app/pdf/PdfReaderActivity.kt

```diff
@@ -0,0 +1,180 @@
+import org.nypl.simplified.books.book_database.api.BookDatabaseException
+import org.nypl.simplified.books.book_database.api.BookDatabaseType
+import org.nypl.simplified.profiles.api.ProfileReadableType
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import java.io.File
+import java.io.InputStream
+
+class PdfReaderActivity : ProfileTimeOutActivity(), PdfFragmentListenerType, TableOfContentsFragmentListenerType {
+
+    companion object {
+        const val TABLE_OF_CONTENTS = "table_of_contents"
+
+        private const val PARAMS_ID = "org.nypl.pdf.android.pdfreader.PdfReaderActivity.params"
+
+        /**
+         * Factory method to start a [PdfReaderActivity]
+         */
+        fun startActivity(
+                from: Activity,
+                parameters: PdfReaderParameters
+        ) {
+
+            val b = Bundle()
+            b.putSerializable(this.PARAMS_ID, parameters)
+            val i = Intent(from, PdfReaderActivity::class.java)
+            i.putExtras(b)
+            from.startActivity(i)
+        }
+    }
+
+    private val log: Logger = LoggerFactory.getLogger(PdfReaderActivity::class.java)
+
+    // vars assigned in onCreate and passed with the intent
+    private lateinit var documentTitle: String
+    private lateinit var pdfFile: File
+    private lateinit var accountId: AccountID
+    private lateinit var id: BookID
+    private lateinit var profile: ProfileReadableType
+    private lateinit var account: AccountType
+    private lateinit var books: BookDatabaseType
+    private lateinit var entry: BookDatabaseEntryType
+    private lateinit var handle: BookDatabaseEntryFormatHandlePDF
+
+    // vars for the activity to pass back to the reader or table of contents fragment
+    private var documentPageIndex: Int = 0
+    private var tableOfContentsList: ArrayList<TableOfContentsItem> = arrayListOf()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        log.debug("onCreate")
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_pdf_reader)
+
+        val intentParams = intent?.getSerializableExtra(PARAMS_ID) as PdfReaderParameters
+        this.documentTitle = intentParams.documentTile
+        this.pdfFile = intentParams.pdfFile
+        this.accountId = intentParams.accountId
+        this.id = intentParams.id
+
+        this.profile = Simplified.getServices()
+                .profilesController
+                .profileCurrent()
+
+        this.account = profile.account(accountId)
+        this.books = account.bookDatabase
+
+        try {
+            this.entry = books.entry(id)
+            this.handle = entry.findFormatHandle(BookDatabaseEntryFormatHandlePDF::class.java)!!
+            this.documentPageIndex = handle.format.lastReadLocation!!
+        } catch (e: BookDatabaseException) {
+            log.error("Could not get lastReadLocation, defaulting to the 1st page")
+            e.printStackTrace()
```

Comment:
Two things here: The three expressions above the catch clause may throw any subtype of `Exception`. Consider what happens if `this.handle == null`...

Secondly, it's better to use the logger you declared above (the line I marked with "Logger!") as then the error would make it to Bugsnag.

```
this.logger.error("could not get last read location: ", e)
```

Reply:
Good catch!